Mathlib/Data/ZMod/Basic.lean

```diff
@@ -167,76 +173,83 @@ theorem cast_eq_val [NeZero n] (a : ZMod n) : (a : R) = a.val := by
 variable {S : Type _} [AddGroupWithOne S]
 
 @[simp]
-theorem Prod.fst_zMod_cast (a : ZMod n) : (a : R × S).fst = a := by cases n <;> simp
-#align prod.fst_zmod_cast Prod.fst_zMod_cast
+theorem _root_.Prod.fst_zmod_cast (a : ZMod n) : (a : R × S).fst = a := by
+  cases n
+  . rfl
+  . simp [ZMod.cast]
+#align prod.fst_zmod_cast Prod.fst_zmod_cast
 
 @[simp]
-theorem Prod.snd_zMod_cast (a : ZMod n) : (a : R × S).snd = a := by cases n <;> simp
-#align prod.snd_zmod_cast Prod.snd_zMod_cast
+theorem _root_.Prod.snd_zmod_cast (a : ZMod n) : (a : R × S).snd = a := by
+  cases n
+  . rfl
+  . simp [ZMod.cast]
+#align prod.snd_zmod_cast Prod.snd_zmod_cast
 
 end
 
-/-- So-named because the coercion is `nat.cast` into `zmod`. For `nat.cast` into an arbitrary ring,
-see `zmod.nat_cast_val`. -/
-theorem nat_cast_zMod_val {n : ℕ} [NeZero n] (a : ZMod n) : (a.val : ZMod n) = a := by
+/-- So-named because the coercion is `Nat.cast` into `ZMod`. For `Nat.cast` into an arbitrary ring,
+see `ZMmod.nat_cast_val`. -/
+theorem nat_cast_zmod_val {n : ℕ} [NeZero n] (a : ZMod n) : (a.val : ZMod n) = a := by
   cases n
   · cases NeZero.ne 0 rfl
   · apply Fin.cast_val_eq_self
-#align zmod.nat_cast_zmod_val ZMod.nat_cast_zMod_val
+#align zmod.nat_cast_zmod_val ZMod.nat_cast_zmod_val
 
-theorem nat_cast_rightInverse [NeZero n] : Function.RightInverse val (coe : ℕ → ZMod n) :=
-  nat_cast_zMod_val
+theorem nat_cast_rightInverse [NeZero n] : Function.RightInverse val ((↑) : ℕ → ZMod n) :=
+  nat_cast_zmod_val
 #align zmod.nat_cast_right_inverse ZMod.nat_cast_rightInverse
 
-theorem nat_cast_zMod_surjective [NeZero n] : Function.Surjective (coe : ℕ → ZMod n) :=
-  nat_cast_rightInverse.Surjective
-#align zmod.nat_cast_zmod_surjective ZMod.nat_cast_zMod_surjective
+theorem nat_cast_zmod_surjective [NeZero n] : Function.Surjective ((↑) : ℕ → ZMod n) :=
+  nat_cast_rightInverse.surjective
+#align zmod.nat_cast_zmod_surjective ZMod.nat_cast_zmod_surjective
 
-/-- So-named because the outer coercion is `int.cast` into `zmod`. For `int.cast` into an arbitrary
-ring, see `zmod.int_cast_cast`. -/
+/-- So-named because the outer coercion is `Int.cast` into `ZMod`. For `Int.cast` into an arbitrary
+ring, see `ZMod.int_cast_cast`. -/
 @[norm_cast]
-theorem int_cast_zMod_cast (a : ZMod n) : ((a : ℤ) : ZMod n) = a := by
+theorem int_cast_zmod_cast (a : ZMod n) : ((a : ℤ) : ZMod n) = a := by
   cases n
-  · rw [Int.cast_id a, Int.cast_id a]
-  · rw [coe_coe, Int.cast_ofNat, Fin.cast_val_eq_self]
-#align zmod.int_cast_zmod_cast ZMod.int_cast_zMod_cast
+  · simp [ZMod.cast, ZMod]
+  · dsimp [ZMod.cast, ZMod]
+    erw [Int.cast_ofNat, Fin.cast_val_eq_self]
```

Comment:
Any idea why the rewrites stopped working?

Reply:
Something to do with needing the fact that `ZMod (n+1) = Fin (n+1)`. The new `rw` unfolds a little less aggressively I guess.